lte/gateway/python/magma/kernsnoopd/handlers.py

```diff
@@ -0,0 +1,170 @@
+
+This source code is licensed under the BSD-style license found in the
+LICENSE file in the root directory of this source tree.
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import abc
+import logging
+from socket import AF_INET, inet_ntop, ntohs
+from struct import pack
+
+import psutil
+from magma.kernsnoopd import metrics
+
+# TASK_COMM_LEN is the string length of binary names that the kernel reports.
+# Value should be the same as found in <linux/sched.h>
+TASK_COMM_LEN = 16
+
+
+class EBPFHandler(abc.ABC):
+    """
+    EBPFHandler class defines the interface for front-end programs
+    corresponding to loaded eBPF programs.
+
+    Method handle() must be implemented by a sub-class. Snooper will call the
+    handle() method of registered front-end programs periodically.
+    """
+
+    def __init__(self, service_registry):
+        self._registry = service_registry
+
+        # only the first TASK_COMM_LEN letters of the service name are relevant
+        # here as the kernel is only sending those in task->comm
+        self._services = [
+            s[:TASK_COMM_LEN] for s in service_registry.list_services()
+        ]
+
+    @abc.abstractmethod
+    def handle(self, bpf) -> None:
+        """
+        Handle() should serve as the entry point of the front-end program
+        performing tasks such as reading metrics collected from the kernel and
+        storing them into Prometheus.
+
+        Args:
+            bpf: the bcc.BPF instance that was used to load the eBPF program
+
+        Raises:
+            NotImplementedError: Implement in sub-class
+        """
+        raise NotImplementedError()
+
+
+class PacketCounter(EBPFHandler):
+    """
+    PacketCounter is the front-end program for ebpf/packet_count.bpf.c
+    """
+
+    def handle(self, bpf):
+        """
+        Handle() reads counters from the loaded packet_count program stored as
+        a dict in 'dest_counters' with key type key_t and value type counter_t
+        defined in ebpf/common.bpf.h
+
+        Args:
+            bpf: bcc.BPF object that was used to load eBPF program into kernel
+        """
+        for key, counters in bpf['dest_counters'].items():
+            d_host = inet_ntop(AF_INET, pack('I', key.daddr))
+            d_port = ntohs(key.dport)
+            service_name = None
+
+            try:
+                # TODO: destination service name inference does not work when
+                # control_proxy is enabled
+                service_name = self._get_source_service(key)
+                # get destination service from host and port
+                dest_service = self._registry.get_service_name(d_host,
+                                                               d_port)
```

Comment:
If two Magma services are communicating, we want to increment MAGMA counters with source and destination service names as labels. This functionality doesn't work yet when traffic is proxied through `control_proxy`.